proposals/0127-get-sysvar-syscall.md

```diff
@@ -0,0 +1,227 @@
+syscalls if the existing design is not ratified.
+
+This tightly-coupled relationship between sysvars and syscalls will perpetually
+swell the syscall interface as changes are made, making maintainence and
+reimplementation difficult.
+
+## Alternatives Considered
+
+Other mechanisms considered that provide sysvar data to BPF programs are as
+follows:
+
+### Sysvar-Specific Syscalls
+
+This approach would involve changing nothing. We'd continue on the same course,
+directly coupling new sysvar fetch functions to new syscalls.
+
+### New VM Memory Regions
+
+A new memory region in the VM could be introduced to store sysvar data, which
+could then be made available to BPF programs through an interface around raw
+pointers, which loads the data similarly to the interface proposed herein.
+
+This approach would depend on direct mapping of host memory to VM memory,
+which has been in development for some time and requires further research
+and development to introduce safely to the VM ABI. This extensive R&D period
+would mean Core BPF program initiatives would remain blocked.
+
+Without direct mapping, sysvar data would be fully copied to this memory
+region each time a VM is created (currently one per transaction processed),
+causing a degradation in runtime performance and likely reducing compute
+by far less compared to the compute savings posed by this proposal.
+
+Even with direct mapping, this change makes address translation more complex
+and therefore slows down _all_ memory accesses.
+
+## New Terminology
+
+N/A.
+
+## Detailed Design
+
+A single `sol_get_sysvar` syscall can be defined, which can be used by all
+forward-facing sysvar interfaces to retrieve data from any sysvar.
+
+```c
+/**
+ * Retrieves a slice of data from a sysvar and copies it to VM memory.
+ *
+ * @param sysvar_id The identifier of the sysvar to retrieve data from.
+ * @param var_addr  VM memory address to copy the retrieved data to.
+ * @param offset    The offset to start copying data from.
+ * @param length    The length of data to copy, in bytes.
+ * @return          A 64-bit unsigned integer error code:
+ *                    - 0 if the operation is successful.
+ *                    - Non-zero error code.
+ *
+ * If the operation is not successful, data will not be written to the
+ * provided VM memory address.
+ */
+uint64_t sol_get_sysvar(
+  /* r1 */ void const * sysvar_id,
+  /* r2 */ uint8_t *    var_addr,
+  /* r3 */ uint64_t     offset,
+  /* r4 */ uint64_t     length,
+);
+
+```
+
+### Control Flow
+
+The syscall aborts the virtual machine if any of these conditions are true:
+
+- Not all bytes in VM memory range `[sysvar_id, sysvar_id + 32)` are readable.
+- Not all bytes in VM memory range `[var_addr, var_addr + length)` are writable.
+- `offset + length` is not in `[0, 2^64)`.
+- `var_addr + length` is not in `[0, 2^64)`.
+- Compute budget is exceeded.
+
+The syscall returns `1` if the sysvar is not present in the Sysvar Cache.
```

Comment:
Its just a list of sysvars. That list of sysvars is specified here. It is confusing to use the term Sysvar Cache when really what is meant is "sysvars supported by this syscall" which is more direct and more explicit given the language in the spec. No one has defined sysvar cache before this SIMD, and this SIMD hardly defines it. No spec or doc provides any color on this either. Lets keep it simple and unambiguous: 

1. the list of supported sysvars is the list here in this SIMD.
2. any new sysvar is included in this list of supported sysvars by default, unless otherwise specified.